test/parallel/test-vm-dynamic-import.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+// Flags: --experimental-vm-modules --experimental-modules
+
+const common = require('../common');
+const assert = require('assert');
+common.crashOnUnhandledRejection();
+
+const vm = require('vm');
+
+async function testModule() {
+  const module = new vm.Module('import("X")');
+  await module.link(common.mustCall((specifier) => {
+    assert.strictEqual(specifier, 'X');
+    return new vm.Module('export const x = 5');
+  }));
+  module.instantiate();
+  const { result } = await module.evaluate();
+  assert.strictEqual((await result).x, 5);
+}
+
+async function testScript() {
+  {
+    let moduleX;
+    const script = new vm.Script('import("vm://X")', {
+      resolveDynamicImport(specifier, scriptOrModule) {
+        if (specifier === 'vm://X') {
+          assert.strictEqual(scriptOrModule, script);
+          return moduleX = new vm.Module('export default import("vm://Y")', { url: 'vm://X' });
+        }
+        if (specifier === 'vm://Y') {
+          assert.strictEqual(scriptOrModule, moduleX);
+          return new vm.Module('export default 5', { url: 'vm://Y' });
+        }
+      },
+    });
```

Comment:
Thanks for clarifying this. So I could have already linked `moduleX` myself with a custom linker, in which case `resolveDynamicImport` does not get called?

So it is only by not linking the module that we use by default the top-level linker?

If so, does this mean that any `import "x"` statements for the `moduleX` instance would be resolved with the same top-level `resolveDynamicImport` method provided to the `vm.Script`?

Finally makes sense, but I need to think about this a little more.

Reply:
linkers are always inherited if not explicitly provided is the rule we came up with before. I like the behaviour but if it's not satisfactory we can try to come up with something else

Reply:
Ah, so it's an extension of Module - Module linker inheritance to Script - Module linkers.

It seems we're kind of caught between two models then. We should try pick a direction clearly here:

A) Call it "linker" on vm.Script, having it act as the inherited linker and only allow it to return a vm.Module instance, or possibly a namespace object as well (just like the linker on Module)
B) Call it "resolveDynamicImport" and only allow it to return a fully executed namespace.

Am I right in thinking a well-defined namespace return for both cases is still blocked by isNamespace? Or is the context security requirement another reason "linker" can't return a namespace object directly?

Reply:
> Am I right in thinking a well-defined namespace return for both cases is still blocked by isNamespace?

yes

> Or is the context security requirement another reason "linker" can't return a namespace object directly?

once you have a namespace there's no worry about contexts, the issue comes from linking modules from separate contexts together.